### DISCUSSION TITLE
vulkanalia vs ash

### DISCUSSION THREAD

**Discussion by HugoPeters1024** [KyleMayes/vulkanalia#204] [General]
the readme already mentioned how the crate is heavily inspired by ash but not why it exists and what the important differences are.

Besides just answering that question for me, I think it should belong in the readme as well.

I do not mean to sound skeptical, I am genuinely interested

**Comment by KyleMayes** [comment]
The following is far from a comprehensive comparison, is to a large degree from memory, and might not reflect the current state of the `ash` project (I haven't really looked at it much since I started this project).

I'll try to flesh this out and link it from the `README.md` in the future.

## The origins of `vulkanalia` (and the tutorial)

`vulkanalia` was created primarily because I wanted to create my own Vulkan bindings crate (yep, just NIH syndrome). I thought it would be a fun and interesting project to work on (which it was!). Beyond that, I looked through the GitHub issues for `ash` and saw some issues I thought I could fix with a fresh approach and rewrite (more on that in the below section with the comparison to `ash`).

I created the Rust port of [vulkan-tutorial.com](https://vulkan-tutorial.com) that also lives in this repository because I thought (a) it would be a good way to teach myself the basics of Vulkan and (b) it would be useful for the Rust community (since at that time there were no Vulkan tutorials using Rust that weren't clearly incomplete). The tutorial weirdly made it to the front-page of HN for an hour or two, so I guess it has helped some people (but then again, pretty much anything with "Rust" in the title has a chance of accomplishing that).

## Comparison to `ash`

Basically I just wanted to do the same thing as `ash` but avoid any prominent issues already encountered by the trailblazers at `ash` (primary examples below). Overall, I'd say most are probably better off with `ash` as this crate doesn't see nearly as much usage so there may be some serious issues lurking in the corners.

https://github.com/ash-rs/ash/issues/91

This crate was designed from the ground up with a split between `vulkanalia-sys` (offering the raw generated Vulkan types) and `vulkanalia` (offering function loading and a thin wrapper around the loaded functions from `vulkanalia-sys`)

https://github.com/ash-rs/ash/issues/344

At the time I started this project, `ash` had many hand-written components, especially for extensions (see https://github.com/ash-rs/ash/issues/178).  My idea was to generate as much of the code in this crate as is reasonable directly from the Vulkan API specification (`vk.xml`) to minimize manual effort in continued maintenance. 

This has largely paid off, I have [a GitHub action that runs daily](https://github.com/KyleMayes/vulkanalia/actions/workflows/update.yml) that automatically opens a PR ([example](https://github.com/KyleMayes/vulkanalia/pull/141)) with any code changes caused by updates to the `vk.xml` file from the official [`Vulkan-Docs` repo](https://github.com/KhronosGroup/Vulkan-Docs/blob/main/xml/vk.xml).

This means that (in theory) this crate can stay up to date with the latest bleeding-edge Vulkan API changes and all I need to do is occasionally merge a PR and cut a release. In reality, sometimes the contents of the `vk.xml` file changes in such a way to break my code generator, but other than needing to fix those issues it has worked pretty well.

It looks like `ash` is making good progress on their own generator rewrite so this advantage may be short-lived.

**Comment by HugoPeters1024** [comment]
Wow thank you so much for the thorough explanation!

**Comment by chuigda** [comment]
One more thing, `ash` works poorly with Jetbrains IDEs which don't use LSP and rust-analyzer. `vulkanalia` is on the contray very friendly to non-LSP based IDEs. Also `vulkanalia-sys`'s generated code is highly human-readable, which can be very useful especially when using a very low-level library.
